impls/src/adapters/mod.rs

```diff
@@ -15,41 +15,80 @@
 mod file;
 mod http;
 mod keybase;
-mod null;
 
-pub use self::file::FileWalletCommAdapter;
-pub use self::http::HTTPWalletCommAdapter;
-pub use self::keybase::KeybaseWalletCommAdapter;
-pub use self::null::NullWalletCommAdapter;
+pub use self::file::PathToSlate;
+pub use self::http::HttpSlateSender;
+pub use self::keybase::{KeybaseAllChannels, KeybaseChannel};
 
 use crate::config::WalletConfig;
-use crate::libwallet::{Error, Slate};
-use std::collections::HashMap;
-
-/// Encapsulate wallet to wallet communication functions
-pub trait WalletCommAdapter {
-	/// Whether this adapter supports sync mode
-	fn supports_sync(&self) -> bool;
+use crate::libwallet::{Error, ErrorKind, Slate};
 
+/// Sends transactions to a corresponding SlateReceiver
+pub trait SlateSender {
 	/// Send a transaction slate to another listening wallet and return result
 	/// TODO: Probably need a slate wrapper type
-	fn send_tx_sync(&self, addr: &str, slate: &Slate) -> Result<Slate, Error>;
-
-	/// Send a transaction asynchronously (result will be returned via the listener)
-	fn send_tx_async(&self, addr: &str, slate: &Slate) -> Result<(), Error>;
-
-	/// Receive a transaction async. (Actually just read it from wherever and return the slate)
-	fn receive_tx_async(&self, params: &str) -> Result<Slate, Error>;
+	fn send_tx(&self, slate: &Slate) -> Result<Slate, Error>;
+}
 
+pub trait SlateReceiver {
 	/// Start a listener, passing received messages to the wallet api directly
 	/// Takes a wallet config for now to avoid needing all sorts of awkward
 	/// type parameters on this trait
 	fn listen(
 		&self,
-		params: HashMap<String, String>,
 		config: WalletConfig,
 		passphrase: &str,
 		account: &str,
 		node_api_secret: Option<String>,
 	) -> Result<(), Error>;
 }
+
+/// Posts slates to be read later by a corresponding getter
+pub trait SlatePutter {
+	/// Send a transaction asynchronously
+	fn put_tx(&self, slate: &Slate) -> Result<(), Error>;
+}
+
+/// Checks for a transaction from a corresponding SlatePutter, returns the transaction if it exists
+pub trait SlateGetter {
+	/// Receive a transaction async. (Actually just read it from wherever and return the slate)
+	fn get_tx(&self) -> Result<Slate, Error>;
+}
+
+/// select a SlateSender based on method and dest fields from, e.g., SendArgs
+pub fn create_sender(method: &str, dest: &str) -> Result<Box<dyn SlateSender>, Error> {
```

Comment:
If this method does not work in the context of I2P or public key based server authentication, it won't be much trouble to simply delete it. It's only used in two places.